drivers/flash/soc_flash_b91.c

```diff
@@ -104,6 +135,21 @@ static int flash_b91_write(const struct device *dev, off_t offset,
 		return -EINVAL;
 	}
 
+	/* need to store data in intermediate RAM buffer in case from flash to flash write */
+	if ((uint32_t)data > FLASH_ORIGIN) {
+
+		buf = k_malloc(len);
+		if (buf == NULL) {
+			return -ENOMEM;
+		}
+
+		/* copy Flash data to RAM */
+		memcpy(buf, data, len);
+
+		/* substitute data with allocated buffer */
+		data = buf;
+	}
+
 	/* take semaphore */
 	if (k_sem_take(&dev_data->write_lock, K_NO_WAIT)) {
 		return -EACCES;
```

Comment:
Should we free the buffer here?

Reply:
Fixed.
Moved k_sem_take before k_malloc.